### PR TITLE
Update usage.md and add an example

### PR DESCRIPTION
**Pull request by hayyp**
If you don't mind though, I think it would be more helpful to also have a simple example file along with the usage document.

### CODE REVIEW

File: doc/usage.md [hunk 1]
```diff
@@ -62,7 +62,7 @@ cothread_t co_active();
 ```
 Return handle to current cothread.
 
-Always returns a valid handle, even when called from the main program thread.
+Always returns a valid handle, even when called from the main program thread. If you call `co_active()` from the main program thread, you will get a handle to a pseudo-coroutine which represents the CPU context of the default thread of execution. Passing this handle to a `co_switch(main_thread_handle)` will lead you back to the main function again, as if a `co_switch(other_thread)` just returns.
```
Comment (Screwtapello) [inline doc/usage.md]: This file is (mostly) wrapped to about 80 columns, this paragraph should be too for consistency.

Comment (hayyp) [reply doc/usage.md]: Sorry, I didn't notice that. It should look better now.

Comment (Screwtapello) [reply doc/usage.md]: Much better, thanks!

**Comment by Screwtapello** [comment]
I like the idea of including an example. Once upon a time libco actually came with an example that was also a benchmark and a smoke-test — perhaps I should dig that code out and add it to the repo. How does that sound?

**Comment by namandixit** [comment]
The language in the description feels a bit "casual", for the lack of a better term. It needs to be precise enough for experts, in addition to being understandable for the beginners.

How does the following look? @ahhzee Is this still understandable to you?

---

## cothread_t
```c
typedef void* cothread_t;
```
Handle to cothread.

Handle must be of type `void*`.

A value of `null` (0) indicates an uninitialized or invalid handle, whereas a non-zero 
value indicates a valid handle. A valid handle is backed by execution state to which the
execution can be `co_switch()`ed to.

## co_active
```c
cothread_t co_active();
```
Return a valid cothread_t.

Note that the handle is valid even if the function is called from a non-cothread
context. To achieve this, we save the execution state in an internal buffer, 
instead of using the user-provided memory. Since this handle is valid, it can
be used to `co_switch` to this context from another cothread. In multi-threaded
applications, make sure to not switch non-cothread context across CPU cores,
to prevent any possible conflicts with the OS scheduler.

**Comment by namandixit** [comment]
> I like the idea of including an example. Once upon a time libco actually came with an example that was also a benchmark and a smoke-test — perhaps I should dig that code out and add it to the repo. How does that sound?

There should probably be two code samples -- one for beginners just trying to get something to compile and run, and another that can serve as both an advanced reference as well as a stress test/benchmark.

**Comment by hayyp** [comment]
> an example that was also a benchmark and a smoke-test

I think that's cool, you can never have too many examples, but I think a simple example can help clarify things and help people get started.

> To achieve this, we save the CPU context in an internal
> buffer, instead of using the user-provided memory. Since this handle is valid, it can
> be used to co_switch to this execution context from another cothread. In multi-threaded
> applications, make sure to not switch non-cothread context across CPU cores,
> to prevent any conflict with OS scheduler.

I think it's less straight-forward but it's fine as long as there is a working example (with proper comments) for them.

**Comment by Screwtapello** [comment]
I dug out the old code that used to come with standalone libco releases, and made PR #29.

In particular, we have:

- `test_args` demonstrates how to spawn a function that takes arguments as a co-routine
- `test_serialization` is an advanced example that demonstrates snapshotting and restoring program state by copying the contents of the coroutine stacks (not safe if your program makes heap allocations at runtime!)
- `test_timing` is the aforementioned example/benchmark/smoke-test

Let me know what you think.

**Comment by hayyp** [comment]
I think they would be very helpful. Maybe I should also move my example to `doc/examples/` and update the usage document? Or should I leave the usage document to @namandixit ?

**Comment by Screwtapello** [comment]
I like the current state of `usage.md` in this PR. I'll merge it.

I think the example in this PR is very similar to `test_timing` in #29, and I'm worried it might be confusing to have both of them. If there's anything those new examples don't cover, I'd love to have your feedback on that PR.

**Comment by Screwtapello** [comment]
Merged in a22bd9dba264e70ee7ea7dc2027af4b076890e43

**Pull request closed without merging**